ocaml/xenopsd/lib/xenops_server.ml

```diff
@@ -1667,10 +1666,18 @@ let rec perform_atomic ~progress_callback ?subtask:_ ?result (op : atomic)
                         )
                 in
                 Some e
-            | Task.Pending _ ->
-                error "Parallel: queue_atomics_and_wait returned a pending task" ;
+            | None | Some (Task.Pending _) ->
+                (* Because pending tasks are filtered out in
+                   queue_atomics_and_wait with task_ended the second case will
+                   never be encountered. The previous boolean used in
```

Comment:
OK I see, so when you get `None`, then the task state would have been Pending.

Also, `event_wait` calls `Xenops_task.check_cancelling task` at the end, which causes an exception to be raised if the task has `cancelling` set to `true`. If fact, this can happen at any time during the 20-minute wait, because of the `~with_cancel:(Xenops_task.with_cancel task)` on `Updates.get`.

So the case here really is only a timeout.